deadlines/src/main/java/com/palantir/deadlines/Deadlines.java

```diff
@@ -179,15 +179,15 @@ private static void checkExpiration(long deadline, boolean internal) {
     static String durationToHeaderValue(long durationNanos) {
         if (durationNanos <= 0) {
             // avoid incorrectly encoding negative values
-            // eventually, checkExpiration will throw before we even get here
             return "0";
         }
         // avoid operations on double and String.format
         return (durationNanos / 1000000000)
                 + "."
                 + ((int) (durationNanos % 1000000000) / 100000000)
                 + ((int) (durationNanos % 100000000) / 10000000)
-                + ((int) (durationNanos % 10000000) / 1000000);
+                // use the ceiling on milliseconds; avoids cutting off requests with very small deadlines
+                + ((int) Math.ceil(durationNanos % 10000000 / 1000000d));
```

Comment:
I don't think this provides the correct value if we provide an input value ending in `999999` since it cannot carry the 1 into the 10-millis position, probably worth verifying with a unit test. We should also test `Long.MAX_VALUE` nanos to make sure we don't overflow _after_ checking `durationNanos` in future refactors.

we could normalize `durationNanos` using something like this:

```java
long ceilingMicrosNanos = durationNanos < (Long.MAX_VALUE - 999999 /*precomputed by the compiler*/)
    ? durationNanos + 999999
    : durationNanos;
```

Reply:
We may also want to call out in a comment that if we're truncating out sub-millisecond precision, a series of requests through the stack which take ~1us would incur a 1ms penalty each, ticking down the clock much faster than we expect. While I don't expect that to impact very many operations, it's much safer to fail open than to fail closed here!

Reply:
alright, i think i've addressed these things now